### PR TITLE
Code Review for ConfigLoader

### PR DESCRIPTION
**Pull request by kprzybyla**
This is a code review for ConfigLoader.

### CODE REVIEW

File: test/foo/lib/config/test_configloader.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+def test_load_from_s3(monkeypatch):
+    reset_config_loader_singleton()
+    monkeypatch.setenv('VAR1', 'not bar')
+
+    mock_file = Mock(autospec='read')
+    mock_file.read.return_value = 'prod'
+    mock_open = Mock()
+    mock_open.return_value = mock_file
+    monkeypatch.setattr('builtins.open', mock_open)
```
[the remark this answers is not in the document]
Comment (kprzybyla) [reply test/foo/lib/config/test_configloader.py:46]: Also, to be honest I haven't used monkeypatching in a long time but shouldn't `unittest.mock.patch` be a better solution here?

File: test/foo/lib/config/test_configloader.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    assert ConfigLoader(defaults={'VAR0': 'default value 0'}) is ConfigLoader()
+
+
+def test_load_from_s3(monkeypatch):
+    reset_config_loader_singleton()
+    monkeypatch.setenv('VAR1', 'not bar')
+
+    mock_file = Mock(autospec='read')
+    mock_file.read.return_value = 'prod'
+    mock_open = Mock()
```
Comment (kprzybyla) [inline test/foo/lib/config/test_configloader.py:44]: We should probably specify the `spec` here.

File: test/foo/lib/config/test_configloader.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    mock_open.return_value = mock_file
+    monkeypatch.setattr('builtins.open', mock_open)
+    mock_s3_loader = Mock()
+    mock_s3_loader.return_value = {'VAR0': 'foo',
+                                   'VAR1': 'bar'}
```
[the remark this answers is not in the document]
Comment (kprzybyla) [reply test/foo/lib/config/test_configloader.py:49]: We could also later update it for the environment variables.

File: test/foo/lib/config/test_configloader.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                              'VAR2': 'foo',
+                              'VAR3': 'not bar'}
+    assert config.source == 'test/foo/lib/config/config.yaml'
+
+
+def test_load_missing():
+    reset_config_loader_singleton()
+    cfg_loader = ConfigLoader()
+    with pytest.raises(ConfigLoaderException):
+        cfg_loader.load('test/foo/lib/config/missing-config.yaml')
```
Comment (kprzybyla) [inline test/foo/lib/config/test_configloader.py:30]: We could also test this error handling for the `env` equal to `prod`. Also, it would be nice to verify that the path is present in the exception message.

File: test/foo/lib/config/test_configloader.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    ConfigLoader._instance = None
+    ConfigLoader._instance_ready = False
+
+
+def test_load(monkeypatch):
+    reset_config_loader_singleton()
+    monkeypatch.setenv('VAR3', 'not bar')
+    cfg_loader = ConfigLoader(defaults={'VAR0': 'default value 0',
+                                        'VAR1': 'default value 1'})
+    config = cfg_loader.load('test/foo/lib/config/config.yaml')
```
Comment (kprzybyla) [inline test/foo/lib/config/test_configloader.py:18]: Those paths works but I could picture that in some specific environment or if someone would like to setup this package in a different manner this might fail. Maybe it is better to use the `__file__` variable and build the absolute path for the configs used inside the tests?

File: test/foo/lib/config/test_configloader.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    assert config._config == {'VAR0': 'fizz',
+                              'VAR1': 'default value 1',
+                              'VAR2': 'foo',
+                              'VAR3': 'not bar'}
+    assert config.source == 'test/foo/lib/config/config.yaml'
```
Comment (kprzybyla) [inline test/foo/lib/config/test_configloader.py:23]: The path is used in multiple places. We should extract it to the variable. Also, usage of `Final` from `typing` for such constants would be probably a good idea.

File: test/foo/lib/config/test_configloader.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
+from unittest.mock import Mock
```
Comment (kprzybyla) [inline test/foo/lib/config/test_configloader.py:1]: It would be probably good to add also the test for using the real S3 bucket.

Comment (kprzybyla) [reply test/foo/lib/config/test_configloader.py:1]: Also, once the implementation is fixed according to the review remarks, we should probably review and think about which things needs to be tested and what kind of tests we should have.

File: foo/lib/config/configloader.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            raise ConfigLoaderException(f'Error when opening {path}')
+        return config
+
+    def load(self, path: str) -> Config:
+        if self._config:
+            return self._config
+        config = self._handler(path)
+        environ = os.environ
+        for key in config:
+            if key in environ:
```
[the remark this answers is not in the document]
Comment (kprzybyla) [reply foo/lib/config/configloader.py:89]: Ok, according to the tests we want to take the environment variable value if it is set instead of the config value. Maybe we should document that in the docstring?

**Review by kprzybyla — COMMENTED** [review]
Overall, we should probably review the requirements and ask the question whether the singleton is really needed here. Because it is the cause of many problems and confusion for the users. Also, the S3 bucket is not tested and this code has a bug inside the implementation for the S3 bucket case. Some tests require a lot of mocking because the functionalities are not split as they should be and the tests needs to be updated to check the corner cases mentioned in the comments.
